homeassistant/components/sensor/gitlab_ci.py

```diff
@@ -0,0 +1,187 @@
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    ATTR_ATTRIBUTION, CONF_SCAN_INTERVAL, CONF_TOKEN, CONF_URL, STATE_UNKNOWN)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+CONF_GITLAB_ID = 'gitlab_id'
+CONF_ATTRIBUTION = "Information provided by https://gitlab.com/"
+
+ICON_HAPPY = 'mdi:emoticon-happy'
+ICON_SAD = 'mdi:emoticon-happy'
+ICON_OTHER = 'mdi:git'
+
+ATTR_BUILD_ID = 'build id'
+ATTR_BUILD_STATUS = 'build_status'
+ATTR_BUILD_STARTED = 'build_started'
+ATTR_BUILD_FINISHED = 'build_finished'
+ATTR_BUILD_DURATION = 'build_duration'
+ATTR_BUILD_COMMIT_ID = 'commit id'
+ATTR_BUILD_COMMIT_DATE = 'commit date'
+ATTR_BUILD_BRANCH = 'master'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_TOKEN): cv.string,
+    vol.Required(CONF_GITLAB_ID): cv.string,
+    vol.Optional(CONF_URL, default='https://gitlab.com'): cv.string,
+    vol.Optional(CONF_SCAN_INTERVAL,
+                 default=timedelta(seconds=300)): cv.time_period,
+})
+
+REQUIREMENTS = ['python-gitlab==1.6.0']
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Sensor platform setup."""
+    _logger = logging.getLogger(__name__)
+
+    _priv_token = config.get(CONF_TOKEN)
+    _gitlab_id = config.get(CONF_GITLAB_ID)
+    _interval = config.get(CONF_SCAN_INTERVAL)
+    _url = config.get(CONF_URL)
+
+    if _priv_token is None:
+        _logger.error('No private access token specified')
+        return False
+    if _gitlab_id is None:
+        _logger.error('No GitLab ID specified')
+        return False
+
+    _gitlab_data = GitLabData(
+        priv_token=_priv_token,
+        gitlab_id=_gitlab_id,
+        interval=_interval,
+        url=_url
+    )
+
+    add_devices([GitLabSensor(_gitlab_id, _priv_token, _gitlab_data)])
+
+
+class GitLabSensor(Entity):
+    """Representation of a Sensor."""
+
+    def __init__(self, gitlab_id, priv_token, gitlab_data):
+        """Initialize the sensor."""
+        self._state = None
+        self._available = False
+        self._gitlab_data = gitlab_data
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return 'GitLab CI Status'
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION,
+            ATTR_BUILD_STATUS: self._status,
+            ATTR_BUILD_STARTED: self._started_at,
+            ATTR_BUILD_FINISHED: self._finished_at,
+            ATTR_BUILD_DURATION: self._duration,
```

Comment:
We should not include duration. You're not allowed to store those in the state machine.

Reply:
Is it being updated while a build is running? If so, it should not be included.

Reply:
No, it's the duration the build took, as reported by the GitLab API. Basically, the API returns data of "This is the build SHA, this is the branch, your CI/CD was successful or failed, and it took X seconds to run". It isn't updated while it's running, it's just a data point returned by the API when the build has completed. It's exactly the same "duration" as the "last_build_duration" reported in the current Travis CI sensor.